### ISSUE TITLE
Equivalent index expression results in 50% perf difference in a triton kernel

### ISSUE THREAD

**Issue by shunting314** [pytorch/pytorch#127581]
### 🐛 Describe the bug

We are debugging another issue https://github.com/pytorch/pytorch/issues/124002 for unpacking int8 tensor. We found the perf is quite bad but a workaround Horace found improves the perf by 50%!.

Comparing the generated kernel, we found actually the slow one and fast one are very similar. The only difference is the slow one uses a less efficient but equivalent index expression.

Slow kernel uses index like: `(5504*x_upper) + (x_lower // 2)`
Fast kernel uses index like: `xindex // 2`

And they are equivalent:
```
tl.device_assert(((5504*x_upper) + (x_lower // 2)) == (xindex // 2), "same index")
```

The perf difference is 50%.
ms_slow=0.165 v.s. ms_fast=0.109

We cound do some index expression optimization in inductor, but this indicates some inefficiency in triton that worth fix.

Full repro: 
https://gist.github.com/shunting314/d3e5f16792579109ee204bc2d5f8f141

cc @ezyang @msaroufim @bdhirsh @anijain2305 @chauhang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @Chillee @eellison @jansel 
cc @htyu can people from Meta triton team take a look?

### Error logs

.

### Minified repro

.

### Versions

,

cc @ezyang @msaroufim @bdhirsh @anijain2305 @chauhang

**Comment by htyu** [comment]
It seems that the only difference between the two kernels are the address computation related logics. We saw address computation affecting performance by increasing register pressure in the past.

This is on A100 or H100? I'm going to take a deeper look.

**Comment by shunting314** [comment]
> This is on A100 or H100? I'm going to take a deeper look.

Thanks! It's on A100

**Comment by htyu** [comment]
So I took a look at the Triton IR and PTX. There are a lot arithmetic operations corresponding to the source changes, which also causes more register pressure. The compiler seems to work expectedly. 

PTX for the fast kernel: https://gist.github.com/htyu/1f780a151fa74f61a215be1c102702a2, 15 registers used.

PTX for the slow kernel: https://gist.github.com/htyu/28ca4088d2db675d101cf750338be42a, 28 registers used.

**Comment by eellison** [comment]
@jansel - you had the idea of doing a faster modulo when inputs were positive. do you remember what that was called ?

**Comment by shunting314** [comment]
@htyu 
I'm still a bit skeptical about this.

Increase register usages from 15 -> 28 seems not a big problem since this is not gonna affect occupancy. Then why it cause 50% perf difference?

Do we see any memory access pattern change in the generated code?

**Comment by shunting314** [comment]
Also is the register usage increase from 15->28 looks reasonable?

**Comment by shunting314** [comment]
@htyu  one thing I found from the ptx file you provided is. I check the load and store for the slow and fast kernel.

They are:

```
        @%p1 ld.global.L1::evict_last.b8 { %rs1 }, [ %rd1 + 0 ];
        @%p4 ld.global.L1::evict_last.b8 { %rs2 }, [ %rd2 + 0 ];
        @%p1 ld.global.L1::evict_last.b8 { %rs3 }, [ %rd3 + 0 ];
        @%p8 ld.global.L1::evict_last.b8 { %rs4 }, [ %rd4 + 0 ];
        @%p1 ld.global.L1::evict_last.b8 { %rs5 }, [ %rd5 + 0 ];
        @%p12 ld.global.L1::evict_last.b8 { %rs6 }, [ %rd6 + 0 ];
        @%p1 ld.global.L1::evict_last.b8 { %rs7 }, [ %rd7 + 0 ];
        @%p16 ld.global.L1::evict_last.b8 { %rs8 }, [ %rd8 + 0 ];
        @%p17 ld.global.L1::evict_last.b8 { %rs9 }, [ %rd1 + 0 ];
        @%p20 ld.global.L1::evict_last.b8 { %rs10 }, [ %rd2 + 0 ];
        @%p17 ld.global.L1::evict_last.b8 { %rs11 }, [ %rd3 + 0 ];
        @%p24 ld.global.L1::evict_last.b8 { %rs12 }, [ %rd4 + 0 ];
        @%p17 ld.global.L1::evict_last.b8 { %rs13 }, [ %rd5 + 0 ];
        @%p28 ld.global.L1::evict_last.b8 { %rs14 }, [ %rd6 + 0 ];
        @%p17 ld.global.L1::evict_last.b8 { %rs15 }, [ %rd7 + 0 ];
        @%p32 ld.global.L1::evict_last.b8 { %rs16 }, [ %rd8 + 0 ];
        @%p1 st.global.v2.b32 [ %rd17 + 0 ], { %r2, %r3 };
```
for the slow kernel.

and 
```
        @%p1 ld.global.L1::evict_last.b8 { %rs1 }, [ %rd1 + 0 ];
        @%p3 ld.global.L1::evict_last.b8 { %rs2 }, [ %rd1 + 0 ];
        @%p1 ld.global.L1::evict_last.b8 { %rs3 }, [ %rd3 + 0 ];
        @%p3 ld.global.L1::evict_last.b8 { %rs4 }, [ %rd3 + 0 ];
        @%p1 ld.global.L1::evict_last.b8 { %rs5 }, [ %rd5 + 0 ];
        @%p3 ld.global.L1::evict_last.b8 { %rs6 }, [ %rd5 + 0 ];
        @%p1 ld.global.L1::evict_last.b8 { %rs7 }, [ %rd7 + 0 ];
        @%p3 ld.global.L1::evict_last.b8 { %rs8 }, [ %rd7 + 0 ];
        @%p17 ld.global.L1::evict_last.b8 { %rs9 }, [ %rd1 + 0 ];
        @%p19 ld.global.L1::evict_last.b8 { %rs10 }, [ %rd1 + 0 ];
        @%p17 ld.global.L1::evict_last.b8 { %rs11 }, [ %rd3 + 0 ];
        @%p19 ld.global.L1::evict_last.b8 { %rs12 }, [ %rd3 + 0 ];
        @%p17 ld.global.L1::evict_last.b8 { %rs13 }, [ %rd5 + 0 ];
        @%p19 ld.global.L1::evict_last.b8 { %rs14 }, [ %rd5 + 0 ];
        @%p17 ld.global.L1::evict_last.b8 { %rs15 }, [ %rd7 + 0 ];
        @%p19 ld.global.L1::evict_last.b8 { %rs16 }, [ %rd7 + 0 ];
        @%p1 st.global.v2.b32 [ %rd17 + 0 ], { %r2, %r3 };
```
for the fast kernel.

We can see that the fast kernel mostly alternating 2 masks while the slow kernel does not. Not sure if this cause the big perf difference

**Comment by shunting314** [comment]
hmm, the memory access pattern seems difference. The slow kernel uses address %rd1 - %rd8 to do the first 8 load; while the fast kernel uses address %rd1, %rd3, %rd5, %rd7 twice for the first 8 load. I feel this is more likely to cause the perf difference

**Comment by htyu** [comment]
Good catch. We can put it in the profiler and see if the kernels are memory-bound.

> Also is the register usage increase from 15->28 looks reasonable?

Yes, as the additional cuda core arithmetics in the slow kernel are tensor-level operations and end up being a sequence of scalar operations in each thread.

**Comment by htyu** [comment]
You think memory coalescing can be an issue?

**Comment by htyu** [comment]
Yeah. BTW, Cuda core resources are really limited (on H100 they are 6% of the chip, the rest 94% are tensor cores) and cuda-core operations should be reduced when possible. We've been working hard on reducing cuda core operations or overlapping them with other operations for other kernels.

**Comment by shunting314** [comment]
> Cuda core resources are really limited

Is there a NCU metric showing the usage of CUDA core (e.g. if it's overloaded)?

**Comment by htyu** [comment]
I usually look at warp stalling on instruction level.

**Comment by htyu** [comment]
So I got the ncu profiles for both kernels. Memory throughput is indeed low for the slow kernel but it seems that the loads are blocked on address computation.


For fast kernel:

<img width="1431" alt="image" src="https://github.com/pytorch/pytorch/assets/15899383/6f62cbb3-7db1-4a8b-afb4-1538aade9bde">


For slow kernel:

<img width="1431" alt="image" src="https://github.com/pytorch/pytorch/assets/15899383/5125555c-9cd9-4668-9723-5a171df35b35">

**Comment by shunting314** [comment]
Thanks @htyu , the graphical profile looks nice (even better than the text one)! 

Do you know if the `xindex = xoffset + tl.arange(0, XBLOCK)[:]` statement translates to the same PTX code in both kernels? Asking since the same line cause 21.92% warp stall in the fast kernel but 58.67% in the slow kernel.

**Comment by htyu** [comment]
> Do you know if the xindex = xoffset + tl.arange(0, XBLOCK)[:] statement translates to the same PTX code in both kernels?

They actually do not. On the fast path, the compiler is a bit smarter that it knows `x_flat = (xindex // 2)` results in the same value for every two loads. Therefore it saves half of the address computation. On the slow path it cannot infer anything, and full computation is needed.
